pkg/ring/lifecycler.go

```diff
@@ -631,9 +868,27 @@ func (i *Lifecycler) changeState(ctx context.Context, state IngesterState) error
 
 	level.Info(util.Logger).Log("msg", "changing ingester state from", "old_state", currState, "new_state", state)
 	i.setState(state)
+
+	// If we're not actively leaving the ring and incremental join is disabled,
+	// we'll set all of our tokens to the same state we just changed into.
+	//
+	// Likewise, we'll do the same if we are actively leaving the ring and
+	// incremental leave is disabled.
+	if !i.cfg.JoinIncrementalTransfer && state != LEAVING {
+		i.setTokensState(state)
+	} else if !i.cfg.LeaveIncrementalTransfer && state == LEAVING {
+		i.setTokensState(state)
+	}
+
 	return i.updateConsul(ctx)
 }
 
+func (i *Lifecycler) getLastRing() *Desc {
+	i.countersLock.Lock()
+	defer i.countersLock.Unlock()
+	return i.lastRing
+}
+
 func (i *Lifecycler) updateCounters(ringDesc *Desc) {
```

Comment:
I'd like to keep `updateCounters` since it's used for caching the results for the number of healthy instances. I'll move updating the last saved ring into another function.

Reply:
It is slightly strange that `lastRing` is protected by `i.countersLock` though :)